VultisigApp/VultisigApp/View Models/CoinSelectionViewModel.swift

```diff
@@ -0,0 +1,210 @@
+            switch coinResult {
+            case .success(let btc):
+                btc.priceProviderId = asset.priceProviderId
+                return btc
+            case .failure(let err):
+                logger.info("fail to get bitcoin address,error:\(err.localizedDescription)")
+            }
+        case .solana:
+            let coinResult = SolanaHelper.getSolana(hexPubKey: vault.pubKeyEdDSA, hexChainCode: vault.hexChainCode)
+            switch coinResult {
+            case .success(let sol):
+                sol.priceProviderId = asset.priceProviderId
+                return sol
+            case .failure(let err):
+                logger.info("fail to get solana address,error:\(err.localizedDescription)")
+            }
+        case .sui:
+            let coinResult = SuiHelper.getSui(hexPubKey: vault.pubKeyEdDSA, hexChainCode: vault.hexChainCode)
+            switch coinResult {
+            case .success(let sui):
+                sui.priceProviderId = asset.priceProviderId
+                return sui
+            case .failure(let err):
+                logger.info("fail to get sui address,error:\(err.localizedDescription)")
+            }
+        case .polkadot:
+            let coinResult = PolkadotHelper.getPolkadot(hexPubKey: vault.pubKeyEdDSA, hexChainCode: vault.hexChainCode)
+            switch coinResult {
+            case .success(let dot):
+                dot.priceProviderId = asset.priceProviderId
+                return dot
+            case .failure(let err):
+                logger.info("fail to get polkadot address,error:\(err.localizedDescription)")
+            }
+        case .gaiaChain:
+            let coinResult = ATOMHelper().getATOMCoin(hexPubKey: vault.pubKeyECDSA, hexChainCode: vault.hexChainCode)
+            switch coinResult {
+            case .success(let atom):
+                atom.priceProviderId = asset.priceProviderId
+                return atom
+            case .failure(let err):
+                logger.info("fail to get solana address,error:\(err.localizedDescription)")
+            }
+        case .kujira:
+            let coinResult = KujiraHelper().getCoin(hexPubKey: vault.pubKeyECDSA, hexChainCode: vault.hexChainCode)
+            switch coinResult {
+            case .success(let kuji):
+                kuji.priceProviderId = asset.priceProviderId
+                return kuji
+            case .failure(let err):
+                logger.info("fail to get solana address,error:\(err.localizedDescription)")
+            }
+        }
+        return nil
+    }
+    private func addToChain(asset: Coin, to vault: Vault) async {
+        do{
+            if let newCoin = getNewCoin(asset: asset, vault: vault) {
+                // Fetch priceProviderId for EVM tokens
+                if !newCoin.isNativeToken, asset.chainType == .EVM {
+                    newCoin.priceProviderId = try await priceService.fetchCoingeckoId(
```

Comment:
failed to save coin to model context The data couldn’t be read because it isn’t in the correct format.
failed to save coin to model context The data couldn’t be read because it isn’t in the correct format.

We can add a thread sleep of 1s to add the tokens if N. It is probably hitting the limit.